New SLU task

siddhu001 opened the pull request.

ftshijt commented on egs2/TEMPLATE/slu1/db.sh:
```diff
@@ -0,0 +1,303 @@
+# Set the path of your corpus
```
Better to use symlink here (to sync the databases with other tasks)

ftshijt commented on egs2/TEMPLATE/slu1/pyscripts/audio/format_wav_scp.py:
```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python3
```
Same for here, symlink would be fine

ftshijt commented on egs2/TEMPLATE/slu1/pyscripts/utils/convert_text_to_phn.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
```
If there is no specific changes, keep using the symlink

ftshijt reviewed (changes requested):
Many thanks for your proposed addition to SLU tasks and it would be very cool to have it standalone. Could you please check the current codebase and try to use symlink as many as possible? It would be a bit difficult to review the code given much duplications.

siddhu001 commented:
Hi @ftshijt thanks a lot for your comments. Yes, I have added sym links to reduce code duplications as much as possible. This PR adds a new SLU task in espnet2 and provides support for training and inference of both baseline and 2-pass SLU systems. 

I have also adapted all of your and @sw005320  suggestions from previous PR https://github.com/espnet/espnet/pull/4261 which includes removing unnecessary and not helpful changes, not changing copy_data_dir.sh(I adapted ST's code for `utt_extra_files` to support transcript file) and using `utils/parse_options.sh` to allow more flexible control on shell command in `run_spm.sh`. I am adding more test cases to increase Codecov coverage now. I am also closing the previous PR https://github.com/espnet/espnet/pull/4261 for now.

ftshijt commented on egs2/slurp/slu1/run.sh:
```diff
@@ -15,10 +15,10 @@ asr_config=conf/tuning/train_asr_bert_conformer_deliberation.yaml
 ./slu_transcript.sh \
     --lang en \
     --ngpu 1 \
+    --stage 11\
```
```suggestion
```

siddhu001 replied:
Done!

ftshijt commented on espnet2/slu/espnet_model.py:
```diff
@@ -461,111 +460,20 @@ def _extract_feats(
-            encoder_out_lens: (Batch,)
-            ys_pad: (Batch, Length)
-            ys_pad_lens: (Batch,)
-        """
```
Expect to have the comments, but you can leave it as TODO for the initial PR.

siddhu001 replied:
Yeah I have added TODO!

ftshijt commented on egs2/TEMPLATE/slu1/slu_transcript.sh:
```diff
@@ -0,0 +1,1657 @@
+#!/usr/bin/env bash
```
Why not make this file as `slu.sh`?

siddhu001 replied:
Good suggestion! I have made it slu.sh

ftshijt commented on egs2/TEMPLATE/slu1/slu_transcript.sh:
```diff
@@ -0,0 +1,1657 @@
+word_vocab_size=10000 # Size of word vocabulary.
+
+# ASR model related
+asr_tag=       # Suffix to the result dir for asr model training.
```
Feel free to change all the `ASR`, `recognition`s in the script

siddhu001 replied:
Yeah I think it will help avoid confusion! I have changed all "asr" references to "slu" references

ftshijt commented on egs2/TEMPLATE/slu1/slu_transcript.sh:
```diff
@@ -0,0 +1,1657 @@
+lm_test_text=    # Text file path of language model evaluation set.
+use_transcript=false # Use transcripts to incorporate sematic information
+two_pass=false # Use transcripts to incorporate sematic information
+pre_postencoder_norm=false
```
This augment seems to be more suitable in model_config, any specific reason for here?

siddhu001 replied:
These variables get used in script. For instance, use_transcript decides if transcript token list need to be created.

ftshijt commented on egs2/TEMPLATE/slu1/slu_transcript.sh:
```diff
@@ -0,0 +1,1657 @@
+                        expand_utt_extra_files="${expand_utt_extra_files} $(basename ${single_file})"
+                    done
+                done
+                echo "${expand_utt_extra_files}"
```
```suggestion
```

siddhu001 replied:
Done!

ftshijt commented on egs2/slurp/slu1/local/data.sh:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env bash
```
Will we replace this version with the old version with ASR? If no, better to use a symlink, if yes, please keep this one.

siddhu001 replied:
Yes, this is different from one in asr because we are also creating transcript. Hence i kept this.

ftshijt commented on egs2/slurp/slu1/local/run_spm.sh:
```diff
@@ -0,0 +1,40 @@
+# This script is called in data preparation step by local/data.sh
```
This seems to be general? How about moving it to scripts/text ?

siddhu001 replied:
Done!

ftshijt commented on egs2/slurp/slu1/local/score.py:
```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env bash
```
Same for here. If it is general, please put those in pyscripts

siddhu001 replied:
Sure I added a symlink to pyscripts.

ftshijt commented on espnet2/bin/slu_inference.py:
```diff
@@ -0,0 +1,700 @@
+from espnet.utils.cli_utils import get_commandline_args
+
+
+class Speech2Text:
```
How about changing it to SpeechUnderstand or some other names than Speech2Text as it is not always true?

siddhu001 replied:
I changed to Speech2Understand

ftshijt commented on espnet2/bin/slu_inference.py:
```diff
@@ -0,0 +1,700 @@
+
+        # 4. Build BeamSearch object
+        if asr_model.use_transducer_decoder:
+            beam_search_transducer = BeamSearchTransducer(
```
Do we need a transducer for the SLU task now? I thought it would be better to simplify the current structure. We can also add back for test, but if we did not test it for now, we could skip it for the initial version.

siddhu001 replied:
I discussed this with Shinji in group meeting last week and we thought it was good idea to keep transducer and interctc right now since they might be useful for future experiments.

ftshijt commented on espnet2/tasks/slu.py:
```diff
@@ -0,0 +1,607 @@
+from espnet2.asr.preencoder.sinc import LightweightSincConvs
+from espnet2.asr.specaug.abs_specaug import AbsSpecAug
+from espnet2.asr.specaug.specaug import SpecAug
+from espnet2.asr.transducer.joint_network import JointNetwork
+from espnet2.asr.transducer.transducer_decoder import TransducerDecoder
```
You can keep it simple by removing the transducer requirements for now

siddhu001 replied:
See above.

ftshijt reviewed (commented):
Very cool update! Some minor suggestions as follow:

siddhu001 commented:
Thanks a lot @ftshijt for all your comments. I have incorporated all your suggestions and also added additional tests to meet Codecov requirements. I believe this PR is ready to be merged @ftshijt @sw005320 . Let me know if you have any other comments.

ftshijt reviewed (approved):
Looks good to go~ Many thanks for all the implementations.

sw005320 commented:
Can you give me a day?
I want to do a quick review.

sw005320 commented on espnet2/bin/slu_inference.py:
```diff
@@ -0,0 +1,702 @@
+        >>> speech2understand = Speech2Understand("slu_config.yml", "slu.pth")
+        >>> audio, rate = soundfile.read("speech.wav")
+        >>> speech2understand(audio)
+        [(text, token, token_int, hypothesis object), ...]
```
The output looks complicated as an API.
Are all of them necessary?

siddhu001 replied:
I tried to make it exactly same as Speech2Text https://github.com/espnet/espnet/blob/f274ebed88e3b4820b23ea71fb7f9f6d56706be9/espnet2/bin/asr_inference.py#L47 so that I can use the same inference API as ASR in Hugging face. Let me know if you still think it needs to be simplified.

sw005320 commented on espnet2/bin/slu_inference.py:
```diff
@@ -0,0 +1,702 @@
+from espnet.utils.cli_utils import get_commandline_args
+
+
+class Speech2Understand:
```
Can we make this class simple if we inherit a class from Speech2Text in ASR?
This is just an idea.
If it complicates your class, we do not have to do it.

siddhu001 replied:
This is a good suggestion! I will take a stab at this and try to simplify espnet2/bin/slu_inference.py, espnet2/slu/espnet_model.py and espnet2/tasks/slu.py using the corresponding ASR files.

sw005320 commented on espnet2/slu/espnet_model.py:
```diff
@@ -0,0 +1,559 @@
+        yield
+
+
+class ESPnetSLUModel(AbsESPnetModel):
```
Ditto.
Can we inherit it from ASRESPnetModel?

sw005320 commented on espnet2/tasks/slu.py:
```diff
@@ -0,0 +1,607 @@
+)
+
+
+class SLUTask(AbsTask):
```
ditt.
Can we use ASRTask?

sw005320 reviewed (commented):
I have some high-level comments.
If it is difficult to deal with it, we can leave it.

siddhu001 commented:
Hi @sw005320, I have used ASRTask in espnet2/tasks/slu.py and ASRESPnetModel in espnet2/slu/espnet_model.py, which have helped avoid some code replication.  espnet2/bin/slu_inference.py was not much simplified using Speech2Text class, so I have kept it as it is. I believe PR is ready to be merged (as all CI checks have passed except 1 that was cancelled).

sw005320 commented:
Cool!

@ftshijt, can we go through the review again (I'll also do it)?
I think this is a very good example of making a new task.

ftshijt commented on espnet2/train/preprocessor.py:
```diff
@@ -169,6 +170,20 @@ def __init__(
                 token_list=token_list,
                 unk_symbol=unk_symbol,
             )
+            if transcript_token_list is not None:
```
It seems to be task-specific. How about we just make it a new class?

siddhu001 replied:
Sure, I have made a new preprocessor class using the CommonPreprocessor class as the base class.

ftshijt reviewed (commented):
Looks great! Just a minor comment to improve the modularization.

sw005320 commented on espnet2/train/preprocessor.py:
```diff
@@ -352,6 +331,88 @@ def __call__(
         return data
 
 
+class CommonPreprocessor_SLU(CommonPreprocessor):
```
Maybe you can rename it to `SLUPreprocessor`?

sw005320 commented:
I think we do not have to increase the coverage, but if you miss some essential functions for the test, please add them. If the current test is good enough, please let me know. Then, I'll merge this PR.

siddhu001 commented:
> I think we do not have to increase the coverage, but if you miss some essential functions for the test, please add them. If the current test is good enough, please let me know. Then, I'll merge this PR.

Yeah I agree. I think the current test is good enough and PR is ready to be merged.

sw005320 commented:
I just merged the PR!
Thanks, @siddhu001!
